include/swift/AST/DiagnosticsSema.def

```diff
@@ -3209,6 +3209,8 @@ ERROR(unsupported_recursive_struct,none,
       "contains it",
       (Type))
 
+WARNING(type_not_constructible,none,
```

Comment:
Since we're only diagnosing direct self-recursion in the payload type, could you make this diagnostic more specific?